Add controllers

SinitsaBogdan opened the pull request.

SinitsaBogdan committed:
create : models Booking

- BookingDto.java
- BookingMapper.java
- Booking.java
- BookingRepository.java
- BookingRepositoryImpl.java
- BookingService.java
- BookingServiceImpl.java

SinitsaBogdan committed:
create : models Item

- ItemDto.java
- ItemMapper.java
- Item.java
- ItemRepository.java
- ItemRepositoryImpl.java
- ItemService.java
- ItemServiceImpl.java

SinitsaBogdan committed:
create : models ItemRequest

- ItemRequestDto.java
- ItemRequestMapper.java
- ItemRequest.java
- ItemRequestRepository.java
- ItemRequestRepositoryImpl.java
- ItemRequestService.java
- ItemRequestServiceImpl.java

SinitsaBogdan committed:
create : models User

- UserDto.java
- UserMapper.java
- User.java
- UserRepository.java
- UserRepositoryImpl.java
- UserService.java
- UserServiceImpl.java

SinitsaBogdan committed:
update : Обновление пакета user - Создание базовых методов контроллера

- UserController.java

SinitsaBogdan committed:
update : Обновление пакета request - Создание базовых методов контроллера

SinitsaBogdan committed:
update : Обновление пакета item - Создание базовых методов контроллера

SinitsaBogdan committed:
update : Обновление пакета booking - Создание базовых методов контроллера

SinitsaBogdan committed:
update : Обновление пакета user.service - Создание базовых методов сервиса

- UserService.java
- UserServiceImpl.java

SinitsaBogdan committed:
update : Обновление пакета request.service - Создание базовых методов сервиса

- ItemRequestService.java
- ItemRequestServiceImpl.java

SinitsaBogdan committed:
update : Обновление пакета item.service - Создание базовых методов сервиса

- ItemService.java
- ItemServiceImpl.java

SinitsaBogdan committed:
update : Обновление пакета booking.service - Создание базовых методов сервиса

- BookingService.java
- BookingServiceImpl.java

SinitsaBogdan committed:
delete : to TODO comment

- BookingController.java
- ItemController.java
- ItemRequestController.java
- UserController.java

SinitsaBogdan committed:
update : Обновление пакета item - Создание базовых моделей

- Item.java
- ItemDto.java

SinitsaBogdan committed:
update : Обновление пакета item.dto - Создание Mapper класса

- ItemMapper.java

SinitsaBogdan committed:
update : pom

SinitsaBogdan committed:
update : Обновление пакета user.dto - Создание Mapper класса

- UserMapper.java

SinitsaBogdan committed:
update : Обновление пакета user - Создание базовых моделей

- UserDto.java
- User.java

SinitsaBogdan committed:
update : user.service - Нейминг методов

- UserService.java
- UserServiceImpl.java

SinitsaBogdan committed:
update : user.repo - Создание базовых методов репозитория

- UserRepository.java
- UserRepositoryImpl.java

SinitsaBogdan committed:
update : request.repo - Создание базовых методов репозитория

- ItemRequestRepository.java
- ItemRequestRepositoryImpl.java

SinitsaBogdan committed:
update : booking.repo - Создание базовых методов репозитория

- BookingRepository.java
- BookingRepositoryImpl.java

SinitsaBogdan committed:
update : item.repo - Создание базовых методов репозитория

- ItemRepository.java
- ItemRepositoryImpl.java

SinitsaBogdan committed:
update : item - Обновление параметров модели

SinitsaBogdan committed:
update : model item.dto

- ItemDto.java

fix : project

- ShareItApp.java

- UserController.java
- UserServiceImpl.java
- UserRepositoryImpl.java

- ItemController.java
- ItemServiceImpl.java
- ItemRepositoryImpl.java

- BookingController.java
- BookingServiceImpl.java
- BookingRepositoryImpl.java.java

- ItemRequestController.java
- ItemRequestServiceImpl.java
- ItemRequestRepositoryImpl.java

SinitsaBogdan committed:
update :  Добавление методов проверки ключа в репозитории

- BookingRepository.java
- BookingRepositoryImpl.java
- ItemRepository.java
- ItemRepositoryImpl.java
- ItemRequestRepository.java
- ItemRequestRepositoryImpl.java
- UserRepository.java
- UserRepositoryImpl.java

SinitsaBogdan committed:
create : HelperClass - Validator

SinitsaBogdan committed:
create : Custom Exception

SinitsaBogdan committed:
update : HelperClass - Validator

SinitsaBogdan committed:
update : ErrorMessage.java

SinitsaBogdan committed:
update : users

- UserDto.java
- UserMapper.java
- UserRepository.java
- UserRepositoryImpl.java
- UserService.java
- UserServiceImpl.java
- UserController.java

SinitsaBogdan committed:
update : users

- UserDto.java
- UserMapper.java
- UserRepository.java
- UserRepositoryImpl.java
- UserService.java
- UserServiceImpl.java
- UserController.java

SinitsaBogdan committed:
update : items

SinitsaBogdan committed:
update : items

SinitsaBogdan committed:
update : Validator.java

SinitsaBogdan committed:
update : Validator.java

SinitsaBogdan committed:
update : ErrorMessage.java

SinitsaBogdan committed:
Merge branch 'add-controllers' of https://github.com/SinitsaBogdan/java-shareit into add-controllers

SinitsaBogdan committed:
update : ErrorMessage.java

SinitsaBogdan committed:
update : ItemRequestController.java

SinitsaBogdan committed:
update : ItemServiceImpl.java

SinitsaBogdan committed:
update : ItemRepositoryImpl.java

SinitsaBogdan committed:
update : ItemRepository.java

SinitsaBogdan committed:
update : ItemController.java

SinitsaBogdan committed:
update : BookingController.java

SinitsaBogdan committed:
refactor : style controller

SinitsaBogdan committed:
update : validator object

- Validator.java
- UserServiceImpl.java
- ItemServiceImpl.java

SinitsaBogdan committed:
Update ShareItApp.java

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,19 @@
+
+@Data
+@Builder
 public class ItemDto {
```
В класс также можно добавить аннотацию через валидации

SinitsaBogdan replied:
Да, я помню что можно подключать аннотации для валидации полей объекта,
но мне принципиально не нравится что пропадает возможность отлова возникающего исключения
и переопределения ответа от сервера.

если нужно переделать, переделаю.

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,19 @@
 public class ItemDto {
+
+    private Long id;
+    private String name, description;
```
На мой взгляд, не очень хорошая практика объявлять переменные через запятую, так как это делает приложение менее гибким. Например, над разными полями могут быть разные аннотации

SinitsaBogdan replied:
поправил

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,19 @@
+    private String name, description;
+
+    @Builder.Default
+    private Boolean available = null;
```
Не обязательно инициализировать поле таким образом, над классом можно повесить аннотацию @JsonInclude(Include.NON_NULL). Таким образом, при сериализации/десериализации все поля c null будут исключены

SinitsaBogdan replied:
да, заменил, про него не подумал.

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/repo/ItemRepositoryImpl.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    public Collection<Item> findAllByUser(Long userId) {
```
В данном случае userId не используется.

SinitsaBogdan replied:
да, данный метод был сделан скорее на бедующее как заготовка.
поправил

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/repo/ItemRepositoryImpl.java:
```diff
@@ -0,0 +1,65 @@
+    public Item update(Item item) {
+
+        Item update = data.get(item.getId());
+        if (item.getName() != null && !item.getName().equals(update.getName())) update.setName(item.getName());
```
Логику проверки на null лучше вынести в сервис, так как это больше похоже на бизнес логику приложения

SinitsaBogdan replied:
перенес логику проверки полей для последующего обновления записи в сервис

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    @Override
+    public List<ItemDto> getAll() {
+        List<ItemDto> result = new ArrayList<>();
+        for (Item item : itemRepository.findAll()) result.add(ItemMapper.mapperItemToDto(item));
```
(Предложение) Хорошо было бы использовать стримы, с помощью метода map можно преобразовывать классы

SinitsaBogdan replied:
переход на функциональный стиль, 
то есть использование стрима все таки в приоритете в реальной разработке ?

переделал

хотя, если рассматривать скрин где условий по больше, 
то использование стрима выглядит более привлекательным
![image](https://github.com/SinitsaBogdan/java-shareit/assets/124495463/23e1b588-4d98-4f6b-9f53-b6f46792688c)

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    @Override
+    public List<ItemDto> getAllByUser(Long userId) {
+        List<ItemDto> result = new ArrayList<>();
+        for (Item item : itemRepository.findAllByUser(userId)) {
```
В данном случае приходят все item, а не только конкретного пользователя. Это можно оптимизировать, если, например, ввести дополнительную мапу(в репозитории) 
Map<Long, List<Item>> userItemIndex 
Она позволит сохранять информацию, какой Item принадлежит какому юзеру

SinitsaBogdan replied:
Добавил новую map, и добавил \ изменил методы для работы с ней

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    public List<ItemDto> getBySearchText(String text) {
+        List<ItemDto> result = new ArrayList<>();
+        if (text.isEmpty()) return result;
+        for (Item item : itemRepository.findAll()) {
```
Было бы хорошо получать уже необходимую информацию из репозитория. С этим может помочь стрим и метод filter

SinitsaBogdan replied:
поправил

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,96 @@
+     * Запрос записи вещи пользователя по ID
+     **/
+    @GetMapping("/{itemId}")
+    public ItemDto getById(
```
В java не принято таким образом делать переносы, так делают например в php. В idea есть автоформатирование, Для этого необходимо пару раз нажать ctr+alt+L

SinitsaBogdan replied:
Отформатировал все контроллеры

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,19 @@
+
+@Data
+@Builder
+public class UserDto {
```
Можно добавить валидацию через аннотации

SinitsaBogdan replied:
Да, я помню что можно подключать аннотации для валидации полей объекта,
но мне принципиально не нравится что пропадает возможность отлова возникающего исключения
и переопределения ответа от сервера.

если нужно переделать, переделаю.

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,19 @@
+ * TODO Sprint add-item-requests.
+ */
+
+@Data
```
также можно добавить @JsonInclude(JsonInclude.Include.NON_NULL) для исключения нулевых полей

SinitsaBogdan replied:
добавил аннотации в модели Dto

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/user/repo/UserRepositoryImpl.java:
```diff
@@ -0,0 +1,72 @@
+
+        User update = data.get(user.getId());
+
+        if (user.getName() != null && !user.getName().equals(update.getName())) update.setName(user.getName());
```
Было бы хорошо проверки на пустые строки вынести в сервис, так как это часть бизнес логики

SinitsaBogdan replied:
поправил

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+    @Override
+    public List<UserDto> getAll() {
+        List<UserDto> result = new ArrayList<>();
+        for (User user : userRepository.findAll()) result.add(UserMapper.mapperUserToDto(user));
```
Можно использовать стримы, метод mup

SinitsaBogdan replied:
поправил

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+    public void delete(Long userId) {
+        Validator.checkIdInUserRepo(userId, userRepository);
+        User user = userRepository.findById(userId);
+        userRepository.deleteById(user);
```
Не совсем логичное название метода, обычно deleteById означает удаление по конкретному id, а не по модели.

SinitsaBogdan replied:
поправил

SinitsaBogdan committed:
fix : project is review

- ItemDto.java
  Обновление полей класса.
  Добавление аннотации JsonInclude

- ItemRepository.java
- ItemRepositoryImpl.java
- ItemServiceImpl.java
  Был удален не используемый метод findAllByUser(Long userId)

- ItemRepositoryImpl.java
- ItemServiceImpl.java
  Логика проверки полей объекта для последующего обновления записи, перемещена в сервис.

- ItemController.java
- BookingController.java
  Автоформатирование кода контроллеров

- UserRepository.java
- UserRepositoryImpl.java
- UserServiceImpl.java
  Обновление логики обновления пользователя

- UserController.java
- UserService.java
- UserServiceImpl.java
  Изменение имени метода delete -> deleteById(Long userId)

- ItemRepository.java
- ItemRepositoryImpl.java
- ItemService.java
- ItemServiceImpl.java
  Перенос части реализации из сервиса в репозиторий.
  Удаление не задействованных методов

SinitsaBogdan committed:
fix : project is review

- ItemDto.java
  Обновление полей класса.
  Добавление аннотации JsonInclude

- ItemRepository.java
- ItemRepositoryImpl.java
- ItemServiceImpl.java
  Был удален не используемый метод findAllByUser(Long userId)

- ItemRepositoryImpl.java
- ItemServiceImpl.java
  Логика проверки полей объекта для последующего обновления записи, перемещена в сервис.

- ItemController.java
- BookingController.java
  Автоформатирование кода контроллеров

- UserRepository.java
- UserRepositoryImpl.java
- UserServiceImpl.java
  Обновление логики обновления пользователя

- UserController.java
- UserService.java
- UserServiceImpl.java
  Изменение имени метода delete -> deleteById(Long userId)

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,87 @@
+    @Override
+    public List<ItemDto> getBySearchText(String text) {
+        if (text.isEmpty()) return new ArrayList<>();
+        return itemRepository.findAll().stream()
```
Было бы хорошо получать из хранилища уже отфильтрованные данные, а не запрашивать их полностью

SinitsaBogdan replied:
перенес часть реализации stream в репозиторий

AlexeyCheuzov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Override
+    public void deleteToUser(Long userId, Long itemId) {
```
Пустой метод в уже работающем коде выглядит не очень хорошо, и не совсем соответствует принципу подстановки Барбары Лисков. Лучше его исключить на данный момент, а добавить в реализацию в будущем

SinitsaBogdan replied:
убрал лишний метод

SinitsaBogdan committed:
Merge branch 'add-controllers' of https://github.com/SinitsaBogdan/java-shareit into add-controllers
